File: webapp/src/lib/render/asset-fetcher.ts
/**
 * Asset Fetcher for downloading audio files from R2 storage.
 * 
 * Handles caching of downloaded files to avoid repeated downloads
 * during the render process.
 */

import * as fs from "fs/promises";
import * as path from "path";
import { createR2ClientFromEnv, R2Client } from "@/lib/r2/client";

export interface AssetFetcherOptions {
  cacheDir?: string;
  tempDir?: string;
  r2Client?: R2Client;
}

export interface DownloadedAsset {
  hashPrefix: string;
  localPath: string;
  contentType: string;
  sizeBytes: number;
  downloadedAt: Date;
}

/**
 * AssetFetcher downloads and caches audio files from R2 storage.
 * 
 * Features:
 * - Downloads audio files from R2 using signed URLs
 * - Caches files locally to avoid repeated downloads
 * - Manages temporary directory for render outputs
 */
export class AssetFetcher {
  private cacheDir: string;
  private tempDir: string;
  private r2Client: R2Client;
  private downloadedAssets: Map<string, DownloadedAsset> = new Map();

  constructor(options: AssetFetcherOptions = {}) {
    this.cacheDir = options.cacheDir ?? "/tmp/sow-assets/cache";
    this.tempDir = options.tempDir ?? "/tmp/sow-assets/temp";
    this.r2Client = options.r2Client ?? createR2ClientFromEnv();
  }

  /**
   * Initialize directories for caching and temp files.
   */
  async initialize(): Promise<void> {
    await fs.mkdir(this.cacheDir, { recursive: true });
    await fs.mkdir(this.tempDir, { recursive: true });
  }

  /**
   * Get the temp directory path.
   */
  async getTempDir(): Promise<string> {
    await fs.mkdir(this.tempDir, { recursive: true });
    return this.tempDir;
  }

  /**
   * Get the cache directory path.
   */
  getCacheDir(): string {
    return this.cacheDir;
  }

  /**
   * Download audio file from R2 and cache locally.
   * 
   * @param hashPrefix - Recording hash prefix
   * @returns Local path to downloaded file or null if failed
   */
  async downloadAudio(hashPrefix: string): Promise<string | null> {
    // Check if already cached
    const cached = this.downloadedAssets.get(hashPrefix);
    if (cached) {
      // Verify file still exists
      try {
        await fs.access(cached.localPath);
        return cached.localPath;
      } catch {
        // File was deleted, remove from cache
        this.downloadedAssets.delete(hashPrefix);
      }
    }

    // Check local cache directory
    const cachePath = path.join(this.cacheDir, `${hashPrefix}.mp3`);
    try {
      await fs.access(cachePath);
      // File exists in cache
      const stats = await fs.stat(cachePath);
      this.downloadedAssets.set(hashPrefix, {
        hashPrefix,
        localPath: cachePath,
        contentType: "audio/mpeg",
        sizeBytes: stats.size,
        downloadedAt: stats.mtime,
      });
      return cachePath;
    } catch {
      // File not in cache, need to download
    }

    try {
      // Generate signed URL
      const signedUrlResult = await this.r2Client.getAudioSignedUrl(hashPrefix, {
        expiresInSeconds: 3600,
      });

      // Download file
      const response = await fetch(signedUrlResult.url);
      if (!response.ok) {
        throw new Error(`Failed to download: ${response.status} ${response.statusText}`);
      }

      // Get file data
      const arrayBuffer = await response.arrayBuffer();
      const buffer = Buffer.from(arrayBuffer);

      // Ensure cache directory exists
      await fs.mkdir(this.cacheDir, { recursive: true });

      // Write to cache
      await fs.writeFile(cachePath, buffer);

      // Record in cache map
      const stats = await fs.stat(cachePath);
      this.downloadedAssets.set(hashPrefix, {
        hashPrefix,
        localPath: cachePath,
        contentType: "audio/mpeg",
        sizeBytes: stats.size,
        downloadedAt: new Date(),
      });

      return cachePath;
    } catch (error) {
      console.error(`Failed to download audio for ${hashPrefix}:`, error);
      return null;
    }
  }

  /**
   * Download LRC file from R2.
   * 
   * @param hashPrefix - Recording hash prefix
   * @returns LRC file content as string or null if failed
   */
  async downloadLrc(hashPrefix: string): Promise<string | null> {
    try {
      // Generate signed URL
      const signedUrlResult = await this.r2Client.getLrcSignedUrl(hashPrefix, {
        expiresInSeconds: 3600,
      });

      // Download file
      const response = await fetch(signedUrlResult.url);
      if (!response.ok) {
        throw new Error(`Failed to download LRC: ${response.status} ${response.statusText}`);
      }

      return await response.text();
    } catch (error) {
      console.error(`Failed to download LRC for ${hashPrefix}:`, error);
      return null;
    }
  }

  /**
   * Get information about a cached asset.
   * 
   * @param hashPrefix - Recording hash prefix
   * @returns Asset info or null if not cached
   */
  getCachedAsset(hashPrefix: string): DownloadedAsset | null {
    return this.downloadedAssets.get(hashPrefix) ?? null;
  }

  /**
   * Check if an asset is cached locally.
   * 
   * @param hashPrefix - Recording hash prefix
   * @returns True if cached
   */
  async isCached(hashPrefix: string): Promise<boolean> {
    // Check memory cache
    if (this.downloadedAssets.has(hashPrefix)) {
      return true;
    }

    // Check filesystem
    const cachePath = path.join(this.cacheDir, `${hashPrefix}.mp3`);
    try {
      await fs.access(cachePath);
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Clear the memory cache (does not delete files).
   */
  clearMemoryCache(): void {
    this.downloadedAssets.clear();
  }

  /**
   * Clear the filesystem cache (deletes cached files).
   */
  async clearFileCache(): Promise<void> {
    try {
      const files = await fs.readdir(this.cacheDir);
      for (const file of files) {
        await fs.unlink(path.join(this.cacheDir, file));
      }
    } catch {
      // Ignore errors
    }
    this.downloadedAssets.clear();
  }

  /**
   * Get cache statistics.
   */
  async getCacheStats(): Promise<{
    fileCount: number;
    totalSizeBytes: number;
  }> {
    try {
      const files = await fs.readdir(this.cacheDir);
      let totalSize = 0;
      for (const file of files) {
        const stats = await fs.stat(path.join(this.cacheDir, file));
        totalSize += stats.size;
      }
      return {
        fileCount: files.length,
        totalSizeBytes: totalSize,
      };
    } catch {
      return {
        fileCount: 0,
        totalSizeBytes: 0,
      };
    }
  }

  /**
   * Clean up temporary files.
   */
  async cleanupTemp(): Promise<void> {
    try {
      const files = await fs.readdir(this.tempDir);
      for (const file of files) {
        await fs.unlink(path.join(this.tempDir, file));
      }
    } catch {
      // Ignore errors
    }
  }
}

File: services/render-worker/src/sow_render_worker/asset_fetcher.py
import logging
import os
import shutil
from pathlib import Path

import urllib3

from sow_render_worker.r2_client import R2Client, create_r2_client_from_env

logger = logging.getLogger(__name__)

DEFAULT_CACHE_DIR = "/tmp/sow-assets/cache"
DEFAULT_TEMP_DIR = "/tmp/sow-assets/temp"
MAX_LRC_SIZE_BYTES = 1024 * 1024


class AssetFetcher:
    def __init__(
        self,
        cache_dir: str | None = None,
        temp_dir: str | None = None,
        r2_client: R2Client | None = None,
    ):
        self._cache_dir = Path(cache_dir or DEFAULT_CACHE_DIR)
        self._temp_dir = Path(temp_dir or DEFAULT_TEMP_DIR)
        self._r2_client = r2_client or create_r2_client_from_env()
        self._http = urllib3.PoolManager(
            timeout=urllib3.Timeout(connect=30, read=300)
        )
        self._lrc_cache: dict[str, str | None] = {}
        self._job_temp_dir: Path | None = None

    def initialize(self) -> None:
        self._cache_dir.mkdir(parents=True, exist_ok=True)
        self._temp_dir.mkdir(parents=True, exist_ok=True)

    def get_temp_dir(self) -> Path:
        self._temp_dir.mkdir(parents=True, exist_ok=True)
        return self._temp_dir

    def get_job_temp_dir(self, job_id: str) -> Path:
        self._job_temp_dir = self._temp_dir / job_id
        self._job_temp_dir.mkdir(parents=True, exist_ok=True)
        return self._job_temp_dir

    def get_cache_dir(self) -> Path:
        return self._cache_dir

    def download_audio(self, hash_prefix: str) -> str | None:
        cache_path = self._cache_dir / f"{hash_prefix}.mp3"
        if cache_path.exists():
            logger.info("Audio cache hit: %s", hash_prefix)
            return str(cache_path)

        try:
            signed_url = self._r2_client.get_audio_signed_url(
                hash_prefix, expires_in_seconds=3600
            )

            response = self._http.request("GET", signed_url, preload_content=False)
            try:
                if response.status != 200:
                    raise RuntimeError(
                        f"Failed to download audio: HTTP {response.status}"
                    )

                self._cache_dir.mkdir(parents=True, exist_ok=True)
                tmp_path = cache_path.with_suffix(".tmp")
                try:
                    total_bytes = 0
                    with open(tmp_path, "wb") as f:
                        for chunk in response.stream(8192):
                            f.write(chunk)
                            total_bytes += len(chunk)
                    os.replace(tmp_path, cache_path)
                except BaseException:
                    try:
                        tmp_path.unlink()
                    except OSError:
                        pass
                    raise
            finally:
                response.release_conn()

            logger.info(
                "Audio downloaded: %s (%d bytes, cached at %s)",
                hash_prefix, total_bytes, cache_path,
            )
            return str(cache_path)
        except Exception as exc:
            logger.exception("Failed to download audio for %s", hash_prefix)
            raise RuntimeError(
                f"Failed to download audio for {hash_prefix}: {exc}"
            ) from exc

    def download_lrc(self, hash_prefix: str) -> str | None:
        if hash_prefix in self._lrc_cache:
            logger.debug("LRC cache hit: %s", hash_prefix)
            return self._lrc_cache[hash_prefix]

        try:
            signed_url = self._r2_client.get_lrc_signed_url(
                hash_prefix, expires_in_seconds=3600
            )

            response = self._http.request("GET", signed_url, preload_content=False)
            chunks: list[bytes] = []
            total_size = 0
            try:
                if response.status != 200:
                    raise RuntimeError(
                        f"Failed to download LRC: HTTP {response.status}"
                    )

                for chunk in response.stream(8192):
                    total_size += len(chunk)
                    if total_size > MAX_LRC_SIZE_BYTES:
                        raise RuntimeError(
                            f"LRC response too large: exceeds {MAX_LRC_SIZE_BYTES} limit"
                        )
                    chunks.append(chunk)
            finally:
                response.release_conn()

            content = b"".join(chunks).decode("utf-8")
            self._lrc_cache[hash_prefix] = content
            logger.info("LRC downloaded: %s (%d bytes)", hash_prefix, total_size)
            return content
        except Exception as exc:
            logger.exception("Failed to download LRC for %s", hash_prefix)
            raise RuntimeError(
                f"Failed to download LRC for {hash_prefix}: {exc}"
            ) from exc

    def cleanup_temp(self) -> None:
        if self._job_temp_dir is not None:
            try:
                shutil.rmtree(self._job_temp_dir, ignore_errors=True)
            except Exception:
                pass
